### PR TITLE
 Add 'mc admin speedtest net' command 

### PR DESCRIPTION
**Pull request by balamurugana**

### CODE REVIEW

**Comment by balamurugana** [comment]
Depends on https://github.com/minio/minio/pull/8112

File: cmd/admin-netperf.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+// netPerfMessage holds network performance information.
+type netPerfMessage struct {
+	Addr    string               `json:"address"`
+	Err     string               `json:"error"`
+	NetPerf []madmin.NetPerfInfo `json:"netPerf"`
+}
+
+func (s netPerfMessage) JSON() string {
+	if len(s.NetPerf) == 0 {
+		s.Err = "network performance not found"
```
Comment (poornas) [inline cmd/admin-netperf.go]:
```suggest
s.Err = "cannot fetch network performance stats"
```

File: cmd/admin-netperf.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	statusJSONBytes, e := json.MarshalIndent(s, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON")
+
+	return string(statusJSONBytes)
+}
+
+func (s netPerfMessage) String() string {
+	if len(s.NetPerf) == 0 {
+		msg := fmt.Sprintf("%s  %s\n", console.Colorize(monitorDegraded, "Source:"), s.Addr) +
+			fmt.Sprintf("   Error: %s\n", console.Colorize(monitorFail, "network performance not found"))
```
Comment (poornas) [inline cmd/admin-netperf.go]:
```suggest
 			fmt.Sprintf("   Error: %s\n", console.Colorize(monitorFail, "cannot fetch network performance stats"))
```

File: cmd/admin-netperf.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	processErr := func(e error) error {
+		switch e.(type) {
+		case *json.SyntaxError:
+			printOfflineErrorMessage(e)
+			return e
+		case *url.Error:
+			printOfflineErrorMessage(e)
+			return e
```
Comment (poornas) [inline cmd/admin-netperf.go]:
```suggest
case *json.SyntaxError:
   fallthrough
case *url.Error:
    printOfflineErrorMessage(e)
    return e
``` 
might be concise

Comment (balamurugana) [reply cmd/admin-netperf.go]: Done

**Comment by Praveenrajmani** [comment]
@balamurugana Travis is failing on - `cmd/admin-netperf.go:62:12: undefined: madmin.NetPerfInfo`

**Comment by poornas** [comment]
@balamurugana , can you rebase this since the minio/minio PR got merged?

**Comment by balamurugana** [comment]
@poornas minio server release is required to update `go.mod`

**Comment by harshavardhana** [comment]
> @poornas minio server release is required to update `go.mod`

you don't need server release @balamurugana you just update by doing `go get github.com/minio/minio/pkg/madmin`  - since the master branch already has the fix. You stick to a commit id for `mc`

**Comment by balamurugana** [comment]
@harshavardhana Does the command fix travis-ci too?

**Comment by harshavardhana** [comment]
> @harshavardhana Does the command fix travis-ci too?

yes that will update your `go.mod` to point to the latest commit of `minio/minio`

**Comment by Praveenrajmani** [comment]
@balamurugana can you update `docs/minio-admin-complete-guide.md`?

File: cmd/admin-netperf.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+func (s netPerfMessage) String() string {
+	if len(s.NetPerf) == 0 {
+		msg := fmt.Sprintf("%s  %s\n", console.Colorize(monitorDegraded, "Source:"), s.Addr) +
+			fmt.Sprintf("   Error: %s\n", console.Colorize(monitorFail, "cannot fetch network performance stats"))
+		return msg
+	}
+
+	msg := fmt.Sprintf("%s  %s\n\n", console.Colorize(monitor, "Source:"), s.Addr)
```
Comment (Praveenrajmani) [inline cmd/admin-netperf.go]: Can we remove this extra `\n`? 

image

**Comment by kannappanr** [comment]
Marking this as blocked until the command line options and commands are approved.

File: cmd/admin-netperf.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		default:
+			// If the error is not nil and unrecognized, just print it and exit
+			fatalIf(probe.NewError(e), "Cannot get service status.")
+		}
+
+		return e
+	}
+
+	// Fetch network performance of all cluster nodes (all MinIO server instances)
+	info, e := client.NetPerfInfo(madmin.DefaultNetPerfSize)
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply cmd/admin-netperf.go]: @harshavardhana Do you need a flag to accept the value or how this needs to be configurable?

[the remark this answers is not in the document]
Comment (harshavardhana) [reply cmd/admin-netperf.go]: Like `--size`

[the remark this answers is not in the document]
Comment (balamurugana) [reply cmd/admin-netperf.go]: Is the flag humanized? If so what unit?

[the remark this answers is not in the document]
Comment (harshavardhana) [reply cmd/admin-netperf.go]: I think you can honor all that humanize.Parse honors..

[the remark this answers is not in the document]
Comment (balamurugana) [reply cmd/admin-netperf.go]: I don't think this is correct. There should be minimum (or maximum) limits else the result is not useful.  My suggestion is to close this with @abperiasamy

[the remark this answers is not in the document]
Comment (harshavardhana) [reply cmd/admin-netperf.go]: Why should there be a maximum value ? Maximum is whatever a network can handle you can't tell that beforehand.. I would suggest we take a value, default can be atleast 1G, sustained large stream will give enough time for switch related issues to show up.

Default value of 100MiB is nothing to test even the tcp buffers for 100Gige networks.

.Of course I will pass it with @abperiasamy .

**Comment by kannappanr** [comment]
Closing this PR. We will be removing `network speed test` from `MinIO` also

**Pull request closed without merging**